c/xudt_rce.c

```diff
@@ -0,0 +1,526 @@
+// current transaction, we consider the extension script to be already
+// validated, no additional check is needed for current extension
+int is_extension_script_validated(mol_seg_t extension_script,
+                                  uint8_t* input_lock_script_hash,
+                                  uint32_t input_lock_script_hash_count) {
+  int err = 0;
+  uint8_t hash[BLAKE2B_BLOCK_SIZE];
+  err = blake2b(hash, BLAKE2B_BLOCK_SIZE, extension_script.ptr,
+                extension_script.size, NULL, 0);
+  CHECK2(err == 0, ERROR_BLAKE2B_ERROR);
+
+  for (uint32_t i = 0; i < input_lock_script_hash_count; i++) {
+    if (memcmp(&input_lock_script_hash[i * BLAKE2B_BLOCK_SIZE], hash,
+               BLAKE2B_BLOCK_SIZE) == 0) {
+      return 0;
+    }
+  }
+  err = ERROR_NOT_VALIDATED;
+exit:
+  return err;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main() {
+#endif
+  // don't move code_buff into global variable. It doesn't work.
+  // it's a ckb-vm bug: the global variable will be freezed:
+  // https://github.com/nervosnetwork/ckb-vm/blob/d43f58d6bf8cc6210721fdcdb6e5ecba513ade0c/src/machine/elf_adaptor.rs#L28-L32
+  // The code can't be loaded into freezed memory.
+  uint8_t code_buff[MAX_CODE_SIZE] __attribute__((aligned(RISCV_PGSIZE)));
+  uint32_t code_used = 0;
+
+  int err = 0;
+  int owner_mode = 0;
+  uint8_t* raw_extension_data = NULL;
+  uint32_t raw_extension_len = 0;
+  XUDTFlags flags = XUDTFlagsPlain;
+  uint8_t
+      input_lock_script_hash[MAX_LOCK_SCRIPT_HASH_COUNT * BLAKE2B_BLOCKBYTES];
+  uint32_t input_lock_script_hash_count = 0;
+  err = parse_args(&owner_mode, &flags, &raw_extension_data, &raw_extension_len,
+                   input_lock_script_hash, &input_lock_script_hash_count);
+  CHECK(err);
+  CHECK2(owner_mode == 1 || owner_mode == 0, ERROR_INVALID_ARGS_FORMAT);
+  if (flags != XUDTFlagsPlain) {
+    CHECK2(raw_extension_data != NULL, ERROR_INVALID_ARGS_FORMAT);
+    CHECK2(raw_extension_len > 0, ERROR_INVALID_ARGS_FORMAT);
+  }
+  err = simple_udt(owner_mode);
+  if (err != 0) {
+    goto exit;
+  }
+
+  if (flags == XUDTFlagsPlain) {
+    err = 0;
+    goto exit;
+  }
+
+  mol_seg_t raw_extension_seg = {0};
+  raw_extension_seg.ptr = raw_extension_data;
+  raw_extension_seg.size = raw_extension_len;
+  CHECK2(MolReader_ScriptVec_verify(&raw_extension_seg, true) == MOL_OK,
+         ERROR_INVALID_ARGS_FORMAT);
+  uint32_t size = MolReader_ScriptVec_length(&raw_extension_seg);
+  for (uint32_t i = 0; i < size; i++) {
+    ValidateFuncType func;
+    mol_seg_res_t res = MolReader_ScriptVec_get(&raw_extension_seg, i);
+    CHECK2(res.errno == 0, ERROR_INVALID_MOL_FORMAT);
+    CHECK2(MolReader_Script_verify(&res.seg, false) == MOL_OK,
+           ERROR_INVALID_MOL_FORMAT);
+
+    mol_seg_t code_hash = MolReader_Script_get_code_hash(&res.seg);
+    mol_seg_t hash_type = MolReader_Script_get_hash_type(&res.seg);
+    mol_seg_t args = MolReader_Script_get_args(&res.seg);
```

Comment:
Call `MolReader_Script_get_args` after `is_extension_script_validated` for tiny optimization. If `input_lock_script_hash` matched, we don't need to extract the args from script.

Reply:
The function "is_extension_script_validated" is not always executed. We can't optimize it.

Reply:
I mean right before `MolReader_Bytes_raw_bytes(&args)`